### PR TITLE
feat(server + client): streaming mutations and queries over HTTP

### LINKED ISSUE #4477: feat: Streaming Mutations / Queries
### Describe the feature you'd like to request

Many upstream APIs (e.g. OpenAI's GPT) stream responses to the client. It would be great if TRPC supported such streaming requests.

### Describe the solution you'd like to see

I'd imagine this would look like a mutation or query where the resolver is an async generator. Example:

```typescript
publicProcedure
   .streamingMutation(async function* ({ input, ctx }) {
     for(let x=0x;<10;x++) {
       yield x;
       await sleep(100);
     }
   });
```

### Describe alternate solutions

An alternative is to set up subscriptions and map the mutation inputs to some identifier in the subscription. The problem with this approach is (a) it depends on websockets, which are difficult to support from an infrastructure perspective, (b) the race conditions between establishing a socket connection and corresponding this with mutations/queries can be miserable.

### Additional information

I'd mostly like to understand any suggestions for how to approach this and whether you'd support a PR. Creating a `streamingMutation` procedure type, for instance, is fairly invasive but may be the correct approach.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4477">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4477/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by KATT**
- Closes #4477
- Closes #4911
- Ref #544 (this PR is not SSE but long-running stuff over HTTP at least)

## 🎯 Changes

- Replace [`httpBatchStreamLink`](https://trpc.io/docs/client/links/httpBatchStreamLink) with an implementation that also handles async generators and deferred promise
- Colocate the serializer and deserialize so it's not spread across `@trpc/client` and `@trpc/server`
- it's basically a simplified variant of [tupleson](https://github.com/trpc/tupleson) that should be easier to fit in with fewer edge-cases

### Other stuff / notes

- I untangled some of the hornets' nest that is httpBatchStream / httpLink / httpBatchStreamLink - the functionality was way too generic and I've made each link a bit more dumb
- Updated docs
  - https://www-git-05-01-stream-trpc.vercel.app/docs/migrate-from-v10-to-v11#reverse-chronological-changelog
  - https://www-git-05-01-stream-trpc.vercel.app/docs/client/links/httpBatchStreamLink#generators
- We should probably add support for generators in subscriptions too (done in #5713)
- Not tested too much - especially around cancellations and stream backpressure etc

### CODE REVIEW

File: packages/server/src/unstable-core-do-not-import/stream/stream.ts [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+  const { data } = opts;
+  let counter = 0 as ChunkIndex;
+  const placeholder = 0 as PlaceholderValue;
+
+  const [stream, controller] = createReadableStream<ChunkData>();
+  const pending = new Set<ChunkIndex>();
+  function maybeClose() {
+    if (pending.size === 0) {
+      controller.close();
+    }
```
Comment (KATT) [inline packages/server/src/unstable-core-do-not-import/stream/stream.ts:137]: I'm not sure if there's a race condition in which this could be called prematurely

**Comment by iamnafets** [comment]
Super, super cool!

Some high-level thoughts from running this in prod:
1) I needed to add some sort of heartbeat which was sent every few seconds and just ignored by the client. This will obviously vary based on the infra you're hosting with, but many load balancers will close the socket if nothing is being sent for a few seconds whereas it wouldn't if you just passed a few bytes.
2) The socket being closed from the client/server was annoying. I needed to race with the socket closed event inside the iterators. There were a ton of edge cases around this which ended up being a pain to debug. Not sure if you've seen them all.

**Comment by iamnafets** [comment]
Another thing I found useful was retaining the `query` and `mutation` methods in the client, but having them instead just use the *last value*. This meant I could re-use the same methods for streaming and non-streaming applications. I instead had `mutateGenerator` and `queryGenerator` for when I explicitly needed to use those.

I don't feel strongly about this, but just a thought.

**Comment by KATT** [comment]
Hey thanks @iamnafets 


> Some high-level thoughts from running this in prod:
> 
> 1. I needed to add some sort of heartbeat which was sent every few seconds and just ignored by the client. This will obviously vary based on the infra you're hosting with, but many load balancers will close the socket if nothing is being sent for a few seconds whereas it wouldn't if you just passed a few bytes.

The idea is not to use queries and mutations for "infinite" iterators - we expect them to stream continuously and then end. What you're describing is a subscription (I think) which is implemented in #5713

> 2. The socket being closed from the client/server was annoying. I needed to race with the socket closed event inside the iterators. There were a ton of edge cases around this which ended up being a pain to debug. Not sure if you've seen them all.

I know. I've seen some, maybe not all 😬 




> Another thing I found useful was retaining the `query` and `mutation` methods in the client, but having them instead just use the _last value_. This meant I could re-use the same methods for streaming and non-streaming applications. I instead had `mutateGenerator` and `queryGenerator` for when I explicitly needed to use those.
> 
> I don't feel strongly about this, but just a thought.

Again, this feels like a subscription and not a query/mutaiton

**Comment by iamnafets** [comment]
> Hey thanks @iamnafets
> 
> > Some high-level thoughts from running this in prod:
> > 
> > 1. I needed to add some sort of heartbeat which was sent every few seconds and just ignored by the client. This will obviously vary based on the infra you're hosting with, but many load balancers will close the socket if nothing is being sent for a few seconds whereas it wouldn't if you just passed a few bytes.
> 
> The idea is not to use queries and mutations for "infinite" iterators - we expect them to stream continuously and then end. What you're describing is a subscription (I think) which is implemented in #5713

Yeah, I’m still talking about finite queries here. We saw hangups happening as soon as 15s which was quite annoying to debug. You can of course just handle this as a user but it’s not exactly a user concern.

**Review by Sheraff — COMMENTED** [review]
Wow that's tough to review. Just a few notes
- I haven't looked at the follow up PR yet at the time of writing this review (https://github.com/trpc/trpc/pull/5713)
- maybe minor but this was initially messing with my head: to me "hydrate" and "dehydrate" would be the opposite of how they're used here. In my mind you *dehydrate* to send over the network, and *hydrate* on the client to "revive" the objects
- as always with these streaming things, I feel like we should test more things but I don't know how to / what exactly. For example
  - does this work correctly in ≠ environments? (for example the `nodeJsStreamToReaderEsque` was missing an error event listener, so maybe we're missing tests for a "node client")
  - does this implementation leak memory? I know with node + promises + streams there are edge-cases where we can have dangling objects without references that still occupy memory, but I'm not really sure how to test for this
- most of this review is nitpicks, the implementation seems solid, and the few tests I tried to write "just to see" all seemed to pass correctly

File: packages/client/src/internals/dataLoader.ts [hunk 2]
```diff
@@ -107,25 +104,30 @@ export function dataLoader<TKey, TValue>(
...
-          for (let i = 0; i < result.length; i++) {
-            const value = result[i]!;
-            unitResolver(i, value);
-          }
+        .then(async (result) => {
+          await Promise.all(
+            result.map(async (valueOrPromise, index) => {
+              const item = batch.items[index]!;
+              try {
+                const value = await Promise.resolve(valueOrPromise);
```
[the remark this answers is not in the document]
Comment (KATT) [reply packages/client/src/internals/dataLoader.ts:118]: Linter complained b/c i was lazy with the types

File: packages/server/src/unstable-core-do-not-import/http/resolveResponse.ts [hunk 3]
```diff
@@ -303,56 +333,55 @@ export async function resolveResponse<TRouter extends AnyRouter>(
...
+           * Not very pretty, but we need to wrap nested data in promises
+           * Our stream producer will only resolve top-level async values or async values that are directly nested in another async value
+           */
+          return {
+            ...response,
+            result: Promise.resolve({
+              ...response.result,
+              data: Promise.resolve(response.result.data),
+            }),
+          };
```
[the remark this answers is not in the document]
Comment (KATT) [reply packages/server/src/unstable-core-do-not-import/http/resolveResponse.ts:367]: Because the "serializer" only looks one level deep of async values

It's smelly

File: www/docs/client/links/httpBatchStreamLink.md [hunk 4]
```diff
@@ -65,10 +63,68 @@ const client = createTRPCClient<AppRouter>({
 
 Compared to a regular `httpBatchLink`, a `unstable_httpBatchStreamLink` will:
 
-- Cause the requests to be sent with a `Trpc-Batch-Mode: stream` header
-- Cause the response to be sent with a `Transfer-Encoding: chunked` and `Vary: trpc-batch-mode` headers
+- Cause the requests to be sent with a `trpc-accept: application/jsonl` header
```
Comment (infodusha) [inline www/docs/client/links/httpBatchStreamLink.md:66]: is it a typo? `application/jsonl`
